src/_pytest/pathlib.py

```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+
+    prepend = "prepend"
+    append = "append"
+    importlib = "importlib"
+
+
+class ImportPathMismatchError(ImportError):
+    """Raised on import_module() if there is a mismatch of __file__'s.
+
+    This can happen when `import_module` is given different filenames with the same basename,
+    when those filenames don't reside in packages.
+    """
+
+
+def import_module(
+    p: Union[str, py.path.local, Path],
+    *,
+    mode: Union[str, ImportMode] = ImportMode.prepend
+) -> ModuleType:
+    """
+    Imports and returns a module from the given path.
+
+    The import mechanism used is controlled by the `mode` parameter:
+
+    * `mode == ImportMode.prepend`: the directory containing the module (or package, taking
+      `__init__.py` files into account) will be put at the *start* of `sys.path`, if not
+      already in  `sys.path`, before being imported with `__import__.
+
+    * `mode == ImportMode.append`: same as `prepend`, but the directory will be appended
+      to the end of `sys.path`.
+
+    * `mode == ImportMode.importlib`: uses more fine control mechanisms provided by `importlib`
+      to import the module, which avoids having to use `__import__` and muck with `sys.path`
+      at all. It effectively allows having same-named test modules in different places.
+
+    :raise ImportPathMismatchError: if after importing the given `path` and the module `__file__`
+        are different. Only raised in `prepend` and `append` modes.
+    """
+    mode = ImportMode(mode)
+
+    path = Path(p)
+
+    if not path.exists():
+        raise ImportError(path)
+
+    if mode == ImportMode.importlib:
+        import importlib.util
+
+        modname = path.stem
+
+        for meta_importer in sys.meta_path:
+            spec = meta_importer.find_spec(modname, [str(path.parent)])
+            if spec is not None:
+                break
+        else:
+            spec = importlib.util.spec_from_file_location(modname, str(path))
+
+        if spec is None:
+            raise ImportError(
+                "Can't find module {} at location {}".format(modname, str(path))
+            )
+        mod = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(mod)  # type: ignore[union-attr]
+        return mod
+
+    pkgpath = resolve_package_path(path)
+    if pkgpath is not None:
+        pkgroot = pkgpath.parent
+        names = str(path.with_suffix("").relative_to(pkgroot)).split(os.path.sep)
+        if names[-1] == "__init__":
+            names.pop()
+        modname = ".".join(names)
+    else:
+        pkgroot = path.parent
+        modname = path.stem
+
+    if mode == ImportMode.append:
+        if str(pkgroot) not in sys.path:
+            sys.path.append(str(pkgroot))
+    else:
+        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
+        if str(pkgroot) != sys.path[0]:
+            sys.path.insert(0, str(pkgroot))
+
+    __import__(modname)
+
```

Comment:
Is it intentional that the `sys.path` is not restored after the `__import__`?

Reply:
Indeed it is; restoring `sys.path` later would cause surprising problems because of delayed imports: for example, a `conftest.py` file imported by this function might have local imports, which would fail at runtime if we restored `sys.path`.

Reply:
Oo that's another good reason to move away from these modes.

A comment with your explanation would be useful!

Reply:
Good point, done! 👍